ddtrace/internal/datadog/profiling/dd_wrapper/src/interface.cpp

```diff
@@ -139,18 +161,26 @@ ddup_init()
         sigaction(SIGSEGV, &(sigaction_handlers), NULL);
 #endif
 
-        g_profile_real[0] = profile_builder.build_ptr();
-        g_profile_real[1] = profile_builder.build_ptr();
-        g_profile = g_profile_real[g_prof_flag];
-        g_uploader = uploader_builder.build_ptr();
-        is_initialized = true;
+        // install the ddup_fork_handler for pthread_atfork
+        // Right now, only do things in the child _after_ fork
+        pthread_atfork(nullptr, nullptr, ddup_fork_handler);
+
+        // Set the global initialization flag
+        is_ddup_initialized = true;
```

Comment:
Does this need to be an atomic write

Reply:
Good question, thank you!

This block is assigned to a variable with [static local storage duration](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables).  The standard guarantees that this section will be initialized only once, and strictly by precisely on thread.  This applies to the body of the lambda as well.  This wasn't your question, and you may know this already, but I wanted to get that piece out of the way just in case.

I don't think it would hurt, but I thought to limit the number of atomic variables for no particular reason.  If this initialization block is running in one thread while _another_ thread is already charging toward an upload, then the upload will deflect.

I didn't think putting the update slightly earlier in the memory pipeline will affect this much--we're not exposing any intermediate states, at least.